components/resource_control/src/resource_group.rs

```diff
@@ -0,0 +1,400 @@
+        }
+    }
+}
+
+pub struct ResourceController {
+    // resource controller name is not used currently.
+    #[allow(dead_code)]
+    name: String,
+    // We handle the priority differently between read and write request:
+    // 1. the priority factor is calcualte based on read/write RU settings.
+    // 2. for read request, we increase a constant virtual time delta at each `get_priority` call
+    //    because the cost can't be calculated at start, so we only increase a constant delta and
+    //    increase the real cost after task is executed; but don't increase it at write because
+    //    the cost is known so we just pre-consume it.
+    is_read: bool,
+    // record consumption of each resource group, name --> resource_group
+    resource_consumptions: DashMap<Vec<u8>, GroupPriorityTracker>,
+    last_min_vt: AtomicU64,
+}
+
+impl ResourceController {
+    pub fn new(name: String, is_read: bool) -> Self {
+        let controller = Self {
+            name,
+            is_read,
+            resource_consumptions: DashMap::new(),
+            last_min_vt: AtomicU64::new(0),
+        };
+        // add the "default" resource group
+        controller.add_resource_group(DEFAULT_RESOURCE_GROUP_NAME.as_bytes().to_owned(), 1);
+        controller
+    }
+
+    fn add_resource_group(&self, name: Vec<u8>, priority_factor: u64) {
+        let vt_delta_for_get = if self.is_read {
+            DEFAULT_PRIORITY_PER_READ_TASK * priority_factor
+        } else {
+            0
+        };
+        let group = GroupPriorityTracker {
+            weight: priority_factor,
+            virtual_time: AtomicU64::new(self.last_min_vt.load(Ordering::Acquire)),
+            vt_delta_for_get,
+        };
+        // maybe update existed group
+        self.resource_consumptions.insert(name, group);
+    }
+
+    fn remove_resource_group(&self, name: &[u8]) {
+        self.resource_consumptions.remove(name);
+    }
+
+    #[inline]
+    fn resource_group(&self, name: &[u8]) -> Ref<'_, Vec<u8>, GroupPriorityTracker> {
+        if let Some(g) = self.resource_consumptions.get(name) {
+            g
+        } else {
+            self.resource_consumptions
+                .get(DEFAULT_RESOURCE_GROUP_NAME.as_bytes())
+                .unwrap()
+        }
+    }
+
+    pub fn consume(&self, name: &[u8], delta: ResourceConsumeType) {
+        self.resource_group(name).consume(delta)
+    }
+
+    pub fn update_min_virtual_time(&self) {
+        let mut min_vt = u64::MAX;
+        let mut max_vt = 0;
+        self.resource_consumptions.iter().for_each(|g| {
+            let vt = g.current_vt();
+            if min_vt > vt {
+                min_vt = vt;
+            }
+            if max_vt < vt {
+                max_vt = vt;
+            }
+        });
+
+        // TODO: use different threshold for different resource type
+        // needn't do update if the virtual different is less than 100ms/100KB.
+        if min_vt + 100_000 >= max_vt {
```

Comment:
why choose 100ms?

Reply:
This is choosed by intuition. As we adjust the vt of all group's each second, we leave some groups' value unchanged if the value is near the biggest value, so we can keep the small schedule advantage between them. This can let the tail lantency a bit stabler.